scheduler/src/cook/hooks.clj

```diff
@@ -0,0 +1,174 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns cook.hooks
+  (:require [clj-time.core :as t]
+            [clj-time.periodic]
+            [chime :as chime]
+            [cook.cache :as ccache]
+            [cook.config :refer [config]]
+            [cook.datomic :as datomic]
+            [cook.hooks-definitions :refer [SchedulerHooks check-job-invocation check-job-submission check-job-submission-default]]
+            [mount.core :as mount]
+            [clojure.tools.logging :as log])
+  (:import (com.google.common.cache Cache CacheBuilder)
+           (java.util.concurrent TimeUnit)))
+
+(def default-accept
+  "A default accept object with an expiration at infinity"
+  {:status :accepted :message "No hook object defined" :cache-expires-at (t/date-time 2783 12 03)})
+
+(def accept-all-hook
+  "A hook object that accepts everything. Available for use as a default hook and unit testing."
+  (reify SchedulerHooks
+    (check-job-submission-default [_] default-accept)
+    (check-job-submission [_ _] default-accept)
+    (check-job-invocation [_ _] default-accept)))
+
+(defn resolve-symbol
+  "Resolve the given symbol to the corresponding Var."
+  [sym]
+  (resolve (some-> sym namespace symbol use) sym))
+
+(defn create-default-hook-object
+  "Returns the hook object. If no hook factory defined, returns an always-accept hook object."
+  [config]
+  (let [{:keys [settings]} config
+        {:keys [hook-factory]} settings
+        {:keys [factory-fn arguments]} hook-factory]
+    (log/info (str "Setting up hooks with factory config: " hook-factory " and factory-fn " factory-fn))
+    (if factory-fn
+      (do
+        (if-let [resolved-fn (resolve-symbol (symbol factory-fn))]
+          (do
+            (log/info (str "Resolved as " resolved-fn " with " arguments))
+            (resolved-fn arguments))
+          (throw (ex-info "Unable to resolve factory function" (assoc hook-factory :ns (namespace factory-fn))))))
+      accept-all-hook)))
+
+;  Contains the hook object that matches to a given job map. This code may create a new hook object or re-use an existing one.
+(mount/defstate hook-object
+  :start (create-default-hook-object config))
+
+(mount/defstate submission-hook-batch-timeout-seconds
```

Comment:
Do we need all these defstates? Can we just add a helper in the config namespace to load these?

Reply:
We discussed and decided to keep it as it is.